servers/src/main/java/tachyon/master/next/IndexSet.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.master.next;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.HashSet;
+import java.lang.reflect.Field;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.Maps;
+
+/**
+ * A set of objects that can be indexed by fields of the object, the field type must be comparable.
+ * In operations that need field name as the parameter, if the object does not have the field, a
+ * RuntimeException will be thrown.
+ */
+public class IndexSet<T> {
+  private Set<T> mSet = new HashSet<T>();
+  /** Map from fieldName to a map from fieldValue to set of T */
+  private Map<String, Map<Object, Set<T>>> mSetIndexedByFieldValue;
+  /** Map to cache the relation from fieldName to Field which is set accessible */
+  private Map<String, Field> mFields;
+
+  /**
+   * Construct a new IndexSet.
+   *
+   * @param fields the field names to index the set of objects
+   */
+  public IndexSet(String... fields) {
```

Comment:
Given that we know the class T, so maybe we shall do a reflection to check if all fields given here are really public fields in class T.